packages/eslint-plugin-wpcalypso/lib/rules/no-package-relative-imports.js

```diff
@@ -0,0 +1,111 @@
+const fs = require( 'fs' );
+const path = require( 'path' );
+
+const getRelativeImports = ( relativeDir ) => {
+	return fs.readdirSync( relativeDir, { withFileTypes: true } ).map( ( file ) => {
+		const fileName = file.name;
+		if ( file.isDirectory() ) return fileName;
+		const extension = path.extname( fileName );
+		if ( extension !== '.js' ) return fileName;
+		return path.basename( fileName, extension );
+	} );
+};
+
+const reportIfPackageRelative = ( { context, node, dependency } ) => {
+	const { mapping, warnOnDynamicImport } = context.options[ 0 ];
+
+	if ( ! dependency ) return;
+	const { value: dependencyName, type: dependencyType } = dependency;
+
+	if ( dependencyType !== 'Literal' ) {
+		if ( warnOnDynamicImport ) {
```

Comment:
Dynamic import usually refers to the async `import( 'modulename' )` syntax which returns a promise, regardless of whether you pass in a literal or a variable. I believe this piece of code is referring to a the latter case specifically, though, where the module name being passed in is not a literal, but rather a variable or an expression of some sort.

If I'm right, perhaps "dynamic import" isn't the best name for this option, since it could lead to confusion? I'm struggling to think of a better alternative, however. Maybe just be literal about it, and call this a "non-literal import"?

Reply:
Agree about the name, it is confusing. Eslint calls "dynamic require" to CJS require with an expression (as opposed to a literal), the same meaning I'm using here: https://github.com/benmosher/eslint-plugin-import/blob/HEAD/docs/rules/no-dynamic-require.md

I'll change it to something else.

Reply:
Not sure about this but I believe `webpack` has stringent rules on what level of dynamism is possible; they still have to be able to resolve the name, meaning you have to be able to provide a glob or some expression in `require.ensure()` (or its modern equivalent)

by this nature though we can't go inspecting the code to reverse-engineer the value used and I support the decision to ignore any of these. if we're not using literals as the import name it's a can of worms that should be caught by PR review anyway

Reply:
For completeness, I only found two cases of this in our codebase:

* https://github.com/Automattic/wp-calypso/blob/master/client/signup/config/step-components.js#L71
* https://github.com/Automattic/wp-calypso/blob/master/client/server/bundler/sections-loader.js#L29

Reply:
Would "programmatic import name" work?